File: mindspore/ops/_op_impl/tbe/reduce_mean.py
# Copyright 2020 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

"""ReduceMean op"""
from mindspore.ops.op_info_register import op_info_register


@op_info_register("""{
    "op_name": "ReduceMean",
    "imply_type": "TBE",
    "fusion_type": "OPAQUE",
    "async_flag": false,
    "binfile_name": "reduce_mean.so",
    "compute_cost": 10,
    "kernel_name": "reduce_mean",
    "partial_flag": true,
    "attr": [
        {
            "name": "axis",
            "param_type": "optional",
            "type": "listInt",
            "value": "all"
        },
        {
            "name": "keep_dims",
            "param_type": "optional",
            "type": "bool",
            "value": "all"
        }
    ],
    "inputs": [
        {
            "index": 0,
            "dtype": [
                "float16","float","float16","int8","uint8"
            ],
            "format": [
                "NC1HWC0","DefaultFormat","DefaultFormat","DefaultFormat","DefaultFormat"
            ],
            "name": "x",
            "need_compile": false,
            "param_type": "required",
            "shape": "all"
        }
    ],
    "outputs": [
        {
            "index": 0,
            "dtype": [
                "float16","float","float16","int8","uint8"
            ],
            "format": [
                "NC1HWC0","DefaultFormat","DefaultFormat","DefaultFormat","DefaultFormat"
            ],
            "name": "y",
            "need_compile": false,
            "param_type": "required",
            "shape": "all"
        }
    ]
}""")
def _reduce_mean_tbe():
    """ReduceMean TBE register"""
    return

File: mindspore/lite/tools/graph_kernel/converter/expanders/reduce_mean.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <memory>

#include "backend/common/graph_kernel/expanders/op_desc_registry.h"
#include "tools/graph_kernel/converter/expanders/activation.h"
#include "mindapi/base/types.h"
#include "ir/dtype.h"

namespace mindspore::graphkernel::expanders {
class ReduceMean : public OpDesc {
 public:
  ReduceMean() {
    std::initializer_list<std::string> attrs{"axis", "keep_dims"};
    (void)validators_.emplace_back(std::make_unique<CheckAttr>(attrs));
  }
  ~ReduceMean() = default;

 protected:
  NodePtrList Expand(const NodePtrList &inputs) override {
    const auto &x = inputs[0];
    auto rank = SizeToLong(x->shape.size());
    auto axis = GetAxisList(attrs_["axis"]);
    (void)std::for_each(axis.begin(), axis.end(), [rank](auto &a) { a = a < 0 ? a + rank : a; });
    if (axis.empty()) {
      for (int64_t i = 0; i < rank; ++i) {
        axis.push_back(i);
      }
    }
    int64_t sz = 1;
    for (size_t i = 0; i < x->shape.size(); ++i) {
      if (std::find(axis.begin(), axis.end(), SizeToLong(i)) != axis.end()) {
        sz *= SizeToLong(x->shape[i]);
      }
    }
    auto sum_x = gb.ReduceSum(x, axis, GetValue<bool>(attrs_["keep_dims"]));
    auto result = gb.Div(sum_x, gb.Const(sz, x->type));
    return {result};
  }
};
EXPANDER_OP_DESC_REGISTER("ReduceMean", ReduceMean);
}  // namespace mindspore::graphkernel::expanders
